IA-1552: account id in urls

beygorghor opened the pull request:
The idea is to add account_id param everywhere in the dashboard in order to debug, share easily links

## Self proof reading checklist

- [ ] Did I use eslint and black formatters
- [ ] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] Did I add translations
- [ ] My migrations file are included
- [ ] Are there enough tests
## Changes

- refactor o `protectedRoute`, main `index` and `App`
- adding accountId in routes, plugin routes, sidebar menu, redirections
- fix tests for new urls


## How to test

Unfortunately you have to pass a bit everywhere.

## Notes

This is working also for anonymous plugins page like embedded calendar

beygorghor committed:
refactor protectedRoutes

beygorghor committed:
break everything

beygorghor committed:
progressing

beygorghor committed:
Merge branch 'main' into IA-1552-account-id

beygorghor committed:
wip on menu

beygorghor committed:
Merge branch 'main' into IA-1552-account-id

beygorghor committed:
fixing redirections

beygorghor committed:
fix cypress tests

beygorghor committed:
Merge branch 'main' into IA-1552-account-id

beygorghor committed:
better

beygorghor commented:
@cypress

beygorghor committed:
fixing js tests

beygorghor commented:
@cypress

beygorghor commented:
@cypress

olethanh commented:
`${siteBaseUrl}/dashboard/orgunits/detail/orgUnitId/${orgUnit.id}/tab/history/accountId/1`

Wouldn't it be more stable and future proof to have the accountId at the start ? Like this ?

`${siteBaseUrl}/dashboard/accountId/1/orgunits/detail/orgUnitId/${orgUnit.id}/tab/history`

would seems less prone to future conflict for me and less chance to mix it with deep linking.

Also wonder if it would be nice to have the account name instead

`${siteBaseUrl}/dashboard/account/pbf_rca/orgunits/detail/orgUnitId/${orgUnit.id}/tab/history`

beygorghor commented:
> `${siteBaseUrl}/dashboard/orgunits/detail/orgUnitId/${orgUnit.id}/tab/history/accountId/1`
> 
> Wouldn't it be more stable and future proof to have the accountId at the start ? Like this ?
> 
> `${siteBaseUrl}/dashboard/accountId/1/orgunits/detail/orgUnitId/${orgUnit.id}/tab/history`
> 
> would seems less prone to future conflict for me and less chance to mix it with deep linking.
> 
> Also wonder if it would be nice to have the account name instead
> 
> `${siteBaseUrl}/dashboard/account/pbf_rca/orgunits/detail/orgUnitId/${orgUnit.id}/tab/history`

@madewulf  ?

olethanh commented:
Also concerning putting it in front, I'm not certain but from my understanding of ReactRouter, if you  put it in front you could just wrap all the current in a parent <Route> component.

I'm seeing they are doing it like this for nesting in their documentation.

```
<Router>
    <Route path="/" component={App}>
      <Route path="about" component={About} />
      <Route path="inbox" component={Inbox}>
        <Route path="messages/:id" component={Message} />
      </Route>
    </Route>
  </Router>
```

https://github.com/remix-run/react-router/blob/v3.2.6/docs/guides/RouteConfiguration.md

quang-le commented on hat/assets/js/apps/Iaso/constants/routes.js:
```diff
@@ -815,4 +815,11 @@ export const routeConfigs = [
+        key: 'accountId',
+    });
+    return newConf;
+});
```
It would save us a loop through the conf to define it directly in each conf, the way we do with `paginationParams`.

quang-le commented on hat/assets/js/apps/Iaso/domains/app/components/MenuItemComponent.js:
```diff
@@ -73,13 +72,19 @@ function MenuItem(props) {
-                target={urlLink ? "_blank" : ""}
+                to={
+                    !hasSubMenu
+                        ? `${fullPath}/accountId/${currentUser.account.id}`
```
We could add the account directly when defining `fullPath`

quang-le commented on hat/assets/js/apps/Iaso/domains/users/components/ProtectedRoute.js:
```diff
@@ -1,109 +1,92 @@
+            ? userHasOneOfPermissions(permissions, currentUser)
+            : true;
+    if (featureFlag && !hasFeatureFlag(currentUser, featureFlag)) {
+        isAuthorized = false;
```
Why not define that earlier and use it in the useEffect above?
Also, line 65, I'm not sure I'm conmfortable with `isAuthorized`  being et to `true` by default. I guess it might be a byproduct of transposing `componentDidUpdate` to `useEffect`. If so, maybe it can be avoided by using `useSkipEffectOnMount` (I don't know, maybe you tried and it didn't work, I'm just thinking out loud really)

quang-le replied:
Ok, I just saw that we set it to `true` in the old code as well. No idea why though

quang-le commented on hat/assets/js/apps/Iaso/domains/users/components/ProtectedRoute.js:
```diff
@@ -1,109 +1,92 @@
+    routeConfig: PropTypes.object.isRequired,
 };
 
-const MapStateToProps = state => ({
-    currentUser: state.users.current,
-});
-
-const MapDispatchToProps = dispatch => ({
-    ...bindActionCreators(
-        {
-            fetchCurrentUser: fetchCurrentUserAction,
-            redirectTo: redirectToAction,
-        },
-        dispatch,
-    ),
-    dispatch,
-});
-
-export default connect(MapStateToProps, MapDispatchToProps)(ProtectedRoute);
```
🍾

quang-le commented on hat/assets/js/apps/Iaso/index.js:
```diff
@@ -29,39 +29,6 @@ export default function iasoApp(
     userHomePage,
 ) {
     const plugins = getPlugins(enabledPluginsName);
-    const allRoutesConfigs = [
-        ...routeConfigs,
-        ...plugins.map(plugin => plugin.routes).flat(),
-    ];
-    const baseRoutes = allRoutesConfigs.map(routeConfig => (
-        <Route
-            path={getPath(routeConfig)}
-            component={
-                routeConfig.allowAnonymous
-                    ? routeConfig.component
-                    : props => (
-                        <ProtectedRoute
-                              {...props}
-                              featureFlag={routeConfig.featureFlag}
-                              permissions={routeConfig.permissions}
-                              component={routeConfig.component(props)}
-                              isRootUrl={routeConfig.isRootUrl}
-                              allRoutes={allRoutesConfigs}
-                          />
-                      )
-            }
-        />
-    ));
-
-    const overrideLandingRoutes = plugins
-        .filter(plugin => plugin.overrideLanding)
-        .map(plugin => plugin.overrideLanding);
-    // using the last plugin override (arbitrary choice)
-    const overrideLanding =
-        overrideLandingRoutes.length > 0
-            ? overrideLandingRoutes[overrideLandingRoutes.length - 1]
```
Out of curiosity: what was gained by moving this code to `App`?

beygorghor replied:
so i could use current user reducer

quang-le commented on hat/assets/js/apps/Iaso/routing/redirections.js:
```diff
@@ -14,73 +14,114 @@ import Page404 from '../components/errors/Page404';
+    ];
 };
+const getRoutes = (baseRoutes, overrideLanding) =>
+    [...baseRoutes].concat(
```
cloneDeep

quang-le commented on hat/assets/js/apps/Iaso/utils/usersUtils.ts:
```diff
@@ -5,6 +5,7 @@ export type Profile = {
     id: string;
     first_name: string;
     user_name: string;
```
this one should be optional as well, no?

quang-le commented on hat/assets/js/apps/Iaso/utils/usersUtils.ts:
```diff
@@ -16,7 +17,7 @@ export type User = {
     first_name: string;
     last_name: string;
     user_name: string;
```
cf supra

quang-le commented on hat/assets/js/apps/Iaso/utils/usersUtils.ts:
```diff
@@ -62,7 +63,7 @@ export const getDisplayName = (
     // Some endpoint have user_name and some username (without the _, fun)
     const userName = user.user_name ?? user?.username;
```
shouldn't it be `user?.user_name` to avoid `undefined.user-name` ?

beygorghor replied:
`user` cannot be null

quang-le commented on hat/assets/js/apps/Iaso/utils/usersUtils.ts:
```diff
@@ -62,7 +63,7 @@ export const getDisplayName = (
-        return userName;
+        return userName || '';
     }
     return `${userName} (${user.first_name ? `${user.first_name}` : ''}${
```
if `userName === ''`,  this will return `(firstName, lastName)`. We should probably add a check to remove the partheses when `!userName`

beygorghor replied:
`userName ` cannot be ''

quang-le reviewed (commented):
Haven't tested the feature, but I'll post already my comments about the code. 
If possible without breaking everything, I'll second @olethanh 's proposal to put the accountId at the beginning of the url rather than the end.

Nice job with `ProtectedRoute`😀!

quang-le commented on hat/assets/js/apps/Iaso/routing/redirections.js:
```diff
@@ -14,73 +14,114 @@ import Page404 from '../components/errors/Page404';
+            to: overrideLanding ?? baseUrls.forms,
+        },
+        {
+            path: `${baseUrls.orgUnits}/accountId/:accountId`,
```
I think the way we redirect now messes up with `goBack` in the router: I get a bug where clicking the back arrow in the `TopBar` brings me back to the url without the accountId then redirected immediately to the same url with the accounId. So I'm basically stuck on the page.

Also, when redirecting without history (deep linking) we need to pass the accountId in the params, otherwise we get a 404 (see Org Unit details for example)

quang-le committed:
Merge remote-tracking branch 'origin' into remove_account_id_from_requests

quang-le committed:
IA-1552: add useSingleTAbleParams

quang-le committed:
IA-1552: remove accountId from GET requests

quang-le committed:
IA-1554: fix back button when deplinking OU detail

quang-le committed:
Fix: add loader to OUType input field (forms)

quang-le committed:
Fix: type error in useGetStorageParams

quang-le committed:
Fix: send correct show_test value for polio campaigns

quang-le reviewed (changes requested):
Works seamlessly overall with the exception of the back button in the `TopBar`:

- When deep linking, user gets a 404

https://user-images.githubusercontent.com/38907762/204509938-62686e5e-c1a2-4689-bf9f-431568c05fb1.mov


- When router has history, users gets stuck on the page. It seems the history remembers the url without the `accountId`, sends the user there, and then redirects again to the page they're trying to leave

https://user-images.githubusercontent.com/38907762/204510923-688211c7-c989-4a71-a952-1391fadffc3a.mov

- Some `GET`requests pass the `accountId` in the params though it's not necessary

I made a branch with a fix for the `GET` requests, a tentative fix for using the back button after deep linking (in OU details), and various minor bug fixes I stumbled upon while reviewing: https://github.com/BLSQ/iaso/tree/remove_account_id_from_requests

All Cypress tests passed locally

beygorghor committed:
code review part 1

beygorghor committed:
Merge branch 'main' into IA-1552-account-id

beygorghor committed:
Merge branch 'main' into IA-1552-account-id

beygorghor commented:
According to our discussion, accound id should be present at the beginning of the url.
We have to update tests and fix @quang-le issues

quang-le committed:
Merge remote-tracking branch 'origin' into IA-1552-account-id

quang-le committed:
Merge branch 'remove_account_id_from_requests' into IA-1552-account-id

quang-le committed:
IA-1552: move accountId at start of url

- move accountId at start of params arrays in routes.js
- reformat url in MenuItemComponent
- Replace redirect with redirectToReplace in ProtectedRoute (bugfix)
- enable redirections to work when omitting accountId

quang-le committed:
IA-1552: add retries to cypress config

- add retries to catch flaky tests

quang-le committed:
IA-1552: update tests (change urls)

- change urls to handle the accountID param
- add timeouts to avoid tests failing

quang-le commented:
I added retries to the cypress tests. the tests will take even more time this way, but it will prevent some flaky tests from failing so it should be an overall win

quang-le committed:
IA-1552: increase timeout on cy request

quang-le committed:
IA-1552: fix test

quang-le reviewed (approved):
I approve my own changes. @beygorghor  I can' set you up as your own reviewer, so I tag you here

quang-le committed:
IA-1552: fix routes for plugins

codeKameleon reviewed (changes requested):
Storages is broken (apart from this page, everything is works well!)
![Capture d’écran du 2022-12-13 16-02-15](https://user-images.githubusercontent.com/25134301/207369021-7551fed8-ccea-4280-b212-75b718e54909.png)

quang-le committed:
IA-1552: fix storage redirection

niconoe reviewed (approved):
I've tried, without success, to generate issues by clicking everywhere for one hour, good job!

Given the nature of the changes, it would still be good if others could do the same!

beygorghor committed:
Merge branch 'main' into IA-1552-account-id

beygorghor committed:
fix cypress test

beygorghor committed:
Merge branch 'main' into IA-1552-account-id
